lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -385,39 +443,29 @@ CloseProximityBuilder::buildFieldReferenceGraph(const Function *F) {
                     dbgs() << "Build partial FRG for BB: " << BB << "\n");
     auto *BBI = FRG->createBasicBlockHelperInfo(&BB);
     for (auto &I : BB) {
-      if (auto FieldNum = StructInfo->getAccessFieldNum(&I)) {
+      Optional<FieldNumType> FieldNum;
+      Optional<ArgNumType> ArgNum;
+      StructInfo->getAccessFieldNumOrArgNum(&I, FieldNum, ArgNum);
+      if (FieldNum.hasValue()) {
         // Case that I is a struct access
         DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs()
                                             << "Found an instruction " << I
                                             << " is a struct access on field ["
                                             << FieldNum.getValue() << "]\n");
         auto *NewNode =
             FRG->createNewNode(FieldNum.getValue(), getMemAccessDataSize(&I));
-        if (BBI->LastNode) {
-          // If there's already a node created in the BB, connect the current
-          // node with previous one
-          DEBUG_WITH_TYPE(DEBUG_TYPE_FRG,
-                          dbgs() << "Previous nodes found in the BB\n");
-          auto C = StructInfo->getExecutionCount(&I).getValue();
-          auto D = BBI->RemainBytes;
-          FRG->connectNodes(BBI->LastNode, NewNode, C, D);
-          BBI->RemainBytes = 0;
-          DEBUG_WITH_TYPE(
-              DEBUG_TYPE_FRG,
-              dbgs() << "Connect new node with previous node in the BB: Field ["
-                     << BBI->LastNode->FieldNum << "] to ["
-                     << FieldNum.getValue() << "] with count: " << C
-                     << " and data " << D << " bytes \n");
-          BBI->LastNode = NewNode;
-        } else {
-          // If there's no previous node, update BBI info
-          DEBUG_WITH_TYPE(
-              DEBUG_TYPE_FRG,
-              dbgs() << "No previous node found. It is the first node\n");
-          BBI->FirstNode = BBI->LastNode = NewNode;
-        }
+        createNewFRGNode(FRG, BBI, NewNode, &I);
```

Comment:
Other than the fact that you can't use auto, it probably makes sense to move the call to lambda outside the if then else. Also, perhaps change the name of the lambda. It looks weird to call FRG->createNode and then call createNewFRGNode